### ISSUE TITLE
PXE boot failing at 'Waiting for link-up on net0'

### ISSUE THREAD

**Issue by victorfong** [RackHD/RackHD#43]
Hi guys, we notice that 1 out of 5 times when we are doing PXE boot, it would fail at the step `Waiting for link-up on net0`. 

The output looks like this:
```
iPXE 1.0.0+ (28ce9b) -- Open Source Network Boot Firmware -- http://ipxe.org
Features: DNS HTTP iSCSI TFTP SRP VLAN AoE ELF MBOOT PXE bzImage COMBOOT Menu PX
EXT
MonoRail iPXE
net0: c0:3f:d5:63:fe:75 using 825791m on PCI00:19.0 (open)
     [Link:down, TX:0 TXE:0 RX:0 RXE:0]
     [Link status: Down (http://ipxe.org/38086101)]
Waiting for link-up on net0 . . . . . . . . . . . . . . . . Down (http://ipxe.org/38086101)

Reboot and Select proper Boot device
```

When this happens, we would need to manually reboot the machine. It's causing automated acceptance test of the RackHD CPI to fail if we don't reboot in time. Is there a way we can make this more stable? The DHCP server is already set to never expire. 

Thanks,
Victor

**Comment by heckj** [comment]
Hey @victorfong, that `down` error (http://ipxe.org/38086101) is most likely specific to your switch and it's configuration. The timing for when that link comes active is a negotiation, and with intelligent (i.e. managed) switches it's controlled by the switches operating system.

If you're using a managed switch, I definite recommend enabling portfast (see http://www.cisco.com/c/en/us/td/docs/switches/lan/catalyst4000/8-2glx/configuration/guide/stp_enha.html for an explanation of the concept from Cisco's world) - an equivalent concept exists in all the major switch vendors.

**Comment by victorfong** [comment]
Hi Joe, I don't think this is a portfast issue. It waits for 15 seconds for the link between the node and the TFTP server to be up. We also saw that every time when the link down error occurs, we see this error in the on-tftp.log:

```
W 2015-12-23T22:16:18.981Z [on-tftp] [Tftp.Server] [Server] Tftp error
 -> /lib/server.js:47
file:          monorail.ipxe
remoteAddress: 172.31.128.36
remotePort:    2070
S 2015-12-23T22:16:19.075Z [on-tftp] [Tftp.Server] [a5dcdd] tftp: 89.229 monorail.ipxe
 -> /lib/server.js:63
```

I wonder if they are related? Thanks.

**Comment by heckj** [comment]
What you're seeing the TFTP logs is serving the monorail.ipxe bootloader, which is what's happening just before the issue you're seeing. The link coming active error is happening from that monorail.ipxe bootloader being loaded and attempting to establish a network connection.

Can you detail how many network connections exist on your target machine, and what's plugged into what? What network switch are you using, what configuration settings are set for it, what is it attached to and is there any spanning tree or related network convergence system running in your environment?

Basically, we need more information to help your troubleshoot and debug this issue....

**Comment by victorfong** [comment]
Sure. The orange box has a build-in "D-Link DGS-1100-16 managed gigabit switch with 802.1q VLAN support". 10 NUCs and the RackHD box are connected to this switch. To isolate this problem, we have tried turning on 1 NUC at a time. I took your suggestion and got into the managed switch, but the 1100-16 switch doesn't have a STP configuration. We have even tried using a different network switch and still saw the problem. 

This is the very last issue in our phase of development. Once we can get this working, our automated CI pipeline will be completed. 

We will be happy to provide whatever information you need. We can do a screen-sharing session with you to describe our setup. Please let us know which configuration file you would like us to share. Thanks a lot for your help!

**Comment by heckj** [comment]
Thanks Victor. A few more questions, and a request for a little work to help the debugging. 

First, Is the DLink switch connected to any other networks, or is this system entirely standalone? If it's uplinked to another switch, how is that configured? Is this a closed layer-2 environment, or linked as layer-2 to a larger broadcast zone?

If you're uplinked/configured to another network, does that network provide DHCP at all, or is it a statically assigned IP address network? Or if you're uplinked, is it to a router that isn't doing layer2 transport? 

Second, you mentioned you tried another switch - could you detail what the other switch was?

Third, does this happen repeatedly and every time, or is this error intermittent - sometimes failing, and sometimes working?

Finally, the suggestions to help debug. The switch data sheet that I found at http://www.dlink.com/-/media/Business_Products/DGS/DGS%201100/Datasheet/DGS_1100_Series_Datasheet_EN_EU.pdf indicates that the switch series has "loop back detection" as well as STP support. Can you determine through their web interface if either of those components are enabled?

When you power off and power on a NUC in this setup, there's a link light that will show up when the network port becomes active. When you power off a NUC, could you verify that the link light does indeed go off (I suspect it may not if you're using the single NIC port for OBM management as well), and when you power it on, could you verify that the link light comes on - and if so, roughly how fast it happens after the NUC is powered on? Instantaneous? A couple of seconds? Longer?

Is there any indicator of port status through the switches web interface that you can look at and that may get updated while you power cycle a NUC? I'd like to know if the switch is reporting the link as down at the same time that the NUC is reporting it's unable to bring up "NET0" per the details at the top of this bug?

**Comment by victorfong** [comment]
Hi Joe,

The switch is totally standalone and not connected to any network.

The other switch we tried is a AirLink 5-Port 1000BASE-T Gigabit Ethernet Switch. We have connected just the RackHD Server and one other NUC to it, but we still saw the error intermittently.

On an average, this link down problem will happen 1/5 times. For the other 4/5 times, the PXE boot can establish the link within a couple of seconds.

In the web interface of the switch, Loopback Detention is presented, but disabled. STP is not presented in the web interface.

There are two lights in the network adapters of each NUC. The green light is always on regardless whether the NUC is powered on or not. The orange light is on only when the NUC is powered on.

For your last question, we will verify that and get back to you.

Thanks,
Victor

**Comment by DavidjohnBlodgett** [comment]
@victorfong @heckj Hi guys, just wanted to circle back on this and verify if this issue has been resolved offline or are we still waiting on information and I should be driving this to completion. thanks!

**Comment by heckj** [comment]
@victorfong good question - are you guys at a good space with debugging and help understanding here, or can we be of further assistance?

**Comment by victorfong** [comment]
@heckj - hi this problem is resolved after we updated the boot image. It can be closed now. Thanks a lot :)

**Comment by benbp** [comment]
@victorfong Perhaps it was https://github.com/RackHD/on-imagebuilder/pull/30 that solved your issue. Glad it's working :)
